### PR TITLE
[GEN] Implement GPU to GEN lowering

### PR DESCRIPTION
**Pull request by FMarno**
Implements the lowering of the GPU dialect to the GEN dialect where possible.
Currently there are only 6 GEN operations, so the lowering is a bit limited.

Signed-off-by: Finlay Marno finlay.marno@codeplay.com

### CODE REVIEW

**Comment by FMarno** [comment]
@etiotto @whitneywhtsang In GPUShuffleToGENLowering, we were a bit unsure what to do with the width argument. GPU to SPIRV lowering checks that the width is constant and equal to the subgroup size, but I don't know if we can check the target information here. Do you have any ideas?

File: mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+public:
+  using OpConversionPattern<GPUOp>::OpConversionPattern;
+  using OpAdaptor = typename GPUOp::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(GPUOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    auto dim = static_cast<std::uint32_t>(adaptor.getDimension());
+    Value idxDim = rewriter.create<arith::ConstantOp>(
+        op->getLoc(), rewriter.getI32IntegerAttr(dim));
```
Comment (victor-eds) [inline mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp:44]:
```suggestion
    Value idxDim = rewriter.create<arith::ConstantIntOp>(op->getLoc(), dim, 32);
```

NIT: I think this should work and is a shortcut for the above. If it works, apply in other places.

**Review by victor-eds — COMMENTED** [review]
Just some minors

File: mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp (hunk 1 above)
Comment (FMarno) [reply mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp]: That works but there was nowhere else to apply it. There is no specialization for a bool that I could find.

File: mlir/include/mlir/Conversion/Passes.td [hunk 2]
```diff
@@ -540,6 +540,16 @@ def LowerHostCodeToLLVMPass : Pass<"lower-host-to-llvm", "ModuleOp"> {
   let dependentDialects = ["LLVM::LLVMDialect"];
 }
 
+//===----------------------------------------------------------------------===//
+// GPUToGEN
+//===----------------------------------------------------------------------===//
+
+def ConvertGpuOpsToGENOps : Pass<"convert-gpu-to-gen"> {
+  let summary = "Generate GEN operations for gpu operations";
+  let constructor = "mlir::createLowerGpuOpsToGENOpsPass()";
```
[the remark this answers is not in the document]
Comment (FMarno) [reply mlir/include/mlir/Conversion/Passes.td]: addressed

File: mlir/include/mlir/Conversion/Passes.h [hunk 3]
```diff
@@ -36,6 +36,7 @@
 #include "mlir/Conversion/GENToLLVM/GENToLLVM.h"
 #include "mlir/Conversion/GENToSPIRV/GENToSPIRV.h"
 #include "mlir/Conversion/GPUCommon/GPUCommonPass.h"
+#include "mlir/Conversion/GPUToGEN/GPUToGEN.h" // TODO should I have the Pass suffix?
```
[the remark this answers is not in the document]
Comment (FMarno) [reply mlir/include/mlir/Conversion/Passes.h]: thanks, I've removed the comment

File: mlir/include/mlir/Conversion/GPUToGEN/GPUToGEN.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+namespace mlir {
+
+template <typename OpT>
+class OperationPass;
+class RewritePatternSet;
+
+#define GEN_PASS_DECL_CONVERTGPUOPSTOGENOPS
+#include "mlir/Conversion/Passes.h.inc"
+
+std::unique_ptr<OperationPass<void>> createLowerGpuOpsToGENOpsPass();
```
[the remark this answers is not in the document]
Comment (FMarno) [reply mlir/include/mlir/Conversion/GPUToGEN/GPUToGEN.h]: I've done that

File: mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    }(gpuMode);
+
+    // TODO unable to validate gpu width parameter, potential for producing
+    // invalid code
+    IntegerAttr widthAttr;
+    if (!matchPattern(adaptor.getWidth(), m_Constant(&widthAttr))) {
+      return rewriter.notifyMatchFailure(
+          op, "shuffle width must be a constant value");
+    }
+
```
[the remark this answers is not in the document]
Comment (FMarno) [reply mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp:93]: Is this not enough https://github.com/intel/llvm/blob/5ea08dbd3e1b0be9be353c03998fa9a851a06fb5/mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp#L86-L87

File: mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp (hunk 1 above)
Comment (victor-eds) [reply mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp]: (bool is just `i1`, but LGTM as is)

File: mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (victor-eds) [reply mlir/lib/Conversion/GPUToGEN/GPUToGEN.cpp:93]: I meant the type for the input `value` and output `shuffleResult` (will be the same). [See](https://mlir.llvm.org/docs/Dialects/GPU/#gpushuffle-gpushuffleop). Our shuffle will not allow all types there. Just put a comment and we can address later.

**Review by victor-eds — APPROVED** [review]

**Comment by whitneywhtsang** [comment]
> @etiotto @whitneywhtsang In GPUShuffleToGENLowering, we were a bit unsure what to do with the width argument. GPU to SPIRV lowering checks that the width is constant and equal to the subgroup size, but I don't know if we can check the target information here. Do you have any ideas?

Target information is not currently available in GEN, we can check if it is one of the expected subgroup sizes (16, 32) for now.

**Comment by victor-eds** [comment]
> Target information is not currently available in GEN, we can check if it is one of the expected subgroup sizes (16, 32) for now.

I'd say we keep this as is till we include a way to encode that in the dialect, i.e., an attribute.

**Comment by FMarno** [comment]
I'm just going to merge this for now, hopefully https://github.com/intel/llvm/pull/13640 will gain traction and a proper solution for the subgroup shuffles can be found.

**Pull request merged**